Mathlib/Tactic/ErwQuestion.lean

```diff
@@ -0,0 +1,104 @@
+-/
+syntax (name := erw?) "erw? " "[" rwRule "]" : tactic
+
+local macro_rules
+  | `(term| verbose $e) => `(term| modify (·.push fun _ => $e))
+
+/--
+Check if two expressions are different at reducible transparency.
+Attempt to log an info message for the first subexpressions which are different
+(but agree at default transparency).
+
+Also returns an array of messages for the `verbose` report.
+-/
+def logDiffs (tk : Syntax) (e₁ e₂ : Expr) : StateT (Array (Unit → MessageData)) MetaM Bool := do
+  withOptions (fun opts => opts.setBool `pp.analyze true) do
+  if ← withReducible (isDefEq e₁ e₂) then
+    verbose m!"{checkEmoji}\n{e₁}\n  and\n{e₂}\nare defeq at reducible transparency."
+    -- They agree at reducible transparency, we're done.
+    return false
+  else
+    verbose m!"{crossEmoji}\n{e₁}\n  and\n{e₂}\nare not defeq at reducible transparency."
+    if ← isDefEq e₁ e₂ then
+      match e₁, e₂ with
+      | Expr.app f₁ a₁, Expr.app f₂ a₂ =>
+        if ← logDiffs tk a₁ a₂ then
+          return true
+        else
+          if ← logDiffs tk f₁ f₂ then
+            return true
+          else
+            logInfoAt tk m!"{crossEmoji}\n{e₁}\n  and\n{e₂}\nare defeq at default transparency, \
+              but not at reducible transparency."
+            return true
+      | Expr.const _ _, Expr.const _ _ =>
+        logInfoAt tk m!"{crossEmoji}\n{e₁}\n  and\n{e₂}\nare defeq at default transparency, \
+            but not at reducible transparency."
+        return true
+      | _, _ =>
+        verbose m!"{crossEmoji}\n{e₁}\n  and\n{e₂}\nare not both applications or constants."
+        return false
+    else
+        verbose m!"{crossEmoji}\n{e₁}\n  and\n{e₂}\nare not defeq at default transparency."
+      return false
+
+elab_rules : tactic
+  | `(tactic| erw?%$tk [$r]) => withMainContext do
+    let verbose := (← getOptions).get `tactic.erw?.verbose (defVal := false)
+    let g ← getMainGoal
+    evalTactic (← `(tactic| erw [$r]))
+    let e := (← instantiateMVars (mkMVar g)).headBeta
+    match e.getAppFnArgs with
+    | (``Eq.mpr, #[_, _, e, _]) =>
+      match e.getAppFnArgs with
+      | (``id, #[ty, e]) =>
+        match ty.eq? with
+        | some (_, tgt, _) =>
+          match (← inferType e).eq? with
+          | some (_, inferred, _) =>
+            let (_, msgs) ← (logDiffs tk tgt inferred).run #[]
+            if verbose then
+              logInfoAt tk <| .joinSep
+                (m!"Expression appearing in target: {tgt}" ::
+                  m!"Expression from `erw`: {inferred}" :: msgs.toList.map (· ())) "\n\n"
+          | _ => logErrorAt tk "Unexpected term produced by `erw`, \
+                  inferred type is not an equality."
+        | _ => logErrorAt tk "Unexpected term produced by `erw`, type hint is not an equality."
+      | _ => logErrorAt tk "Unexpected term produced by `erw`, not of the form: `Eq.mpr (id _) _`."
+    | _ => logErrorAt tk "Unexpected term produced by `erw`, head is not an `Eq.mpr`."
```

Comment:
This nesting would be lot clearer with `let $patt := _ | logErrorAt tk _`

Reply:
Indeed, that is *way* better.